### PR TITLE
[IMP] account: Improve perf of reconciliation

### PR DESCRIPTION
**Pull request by smetl**
Benchmark:
A batch of 1000 payments and a move of 1000 lines to reconcile each payment separately.

Current observed time: +-288s

Improvement:
Adds some context keys when creating the account.full.reconcile to avoid the recomputation of dynamic lines (invoices), synchronization of payments/statement lines and check if the move is well balanced. => +-288s => +-34s

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move_line.py [hunk 1]
```diff
@@ -2395,11 +2395,18 @@ def is_line_reconciled(line, has_multiple_currencies):
...
+                .with_context(
+                    skip_invoice_sync=True,
+                    skip_invoice_line_sync=True,
+                    skip_account_move_synchronization=True,
+                    check_move_validity=False,
+                ) \
+                .create({
+                    'exchange_move_id': exchange_move and exchange_move.id,
+                    'partial_reconcile_ids': [(6, 0, involved_partials.ids)],
+                    'reconciled_line_ids': [(6, 0, involved_lines.ids)],
```
Comment (Levizar) [inline addons/account/models/account_move_line.py:2408]: use the `Command.set`

File: addons/account/models/account_move.py [hunk 2]
```diff
@@ -3569,8 +3569,29 @@ def button_set_checked(self):
...
+                    UNION ALL
+
+                    SELECT exchange_move_id
+                    FROM account_partial_reconcile
+                    WHERE exchange_move_id IN %s
+                """,
+                [tuple(self.ids), tuple(self.ids)],
+            )
+            for row in self._cr.fetchall():
+                exchange_move_ids.add(row[0])
```
Comment (Levizar) [inline addons/account/models/account_move.py:3591]: Can't you just set a `SELECT DISTINCT exchange_move_id FROM (subquery)` so we don't need to create a set for this.

File: addons/account/tests/test_account_move_entry.py [hunk 3]
```diff
@@ -857,3 +857,53 @@ def test_misc_tax_autobalance(self):
...
+        moves.action_post()
+
+        res = moves.line_ids\
+            .filtered(lambda x: x.account_id == self.company_data['default_account_receivable'])\
+            .reconcile()
+
+        self.assertTrue(res.get('partials'))
+        exchange_diff = res['partials'].exchange_move_id
+        self.assertTrue(exchange_diff)
+        with self.assertRaises(UserError), self.cr.savepoint():
```
Comment (Levizar) [inline addons/account/tests/test_account_move_entry.py:908]: I think we don't need the `savepoint` here

File: addons/account/models/account_bank_statement_line.py [hunk 4]
```diff
@@ -661,12 +661,29 @@ def _retrieve_partner(self):
...
+                [
+                    ('parent_id', '=', False),
+                    ('name', 'ilike', self.partner_name),
+                    ('company_id', '=', self.company_id.id),
+                ],
+                [
+                    ('parent_id', '=', False),
+                    ('name', 'ilike', self.partner_name),
+                    ('company_id', '=', False),
+                ],
```
Comment (Levizar) [inline addons/account/models/account_bank_statement_line.py:684]: I don't understand why it repeats twice

**Review by Levizar — CHANGES_REQUESTED** [review]

File: addons/account/models/account_bank_statement_line.py (hunk 4 above)
Comment (Levizar) [reply addons/account/models/account_bank_statement_line.py:684]: The diff is the `ilike`.
Once it's `=ilike` ---> exact matching
then it's `ilike` ---> with %stuff%

**Comment by smetl** [comment]
@Levizar r+ plz :D

**Comment by Levizar** [comment]
@robodoo rebase-ff r+

**Comment by Levizar** [comment]
I just removed the trailing space in the query

**Comment by Levizar** [comment]
@robodoo rebase-ff r+

**Comment by smetl** [comment]
@Levizar Thank you! :D

**Comment by Levizar** [comment]
@smetl ;)

**Pull request closed without merging**